src/host.rs

```diff
@@ -88,6 +132,159 @@ impl Udp {
+
+        trace!(?addr, protocol = %"UDP", "Unbind");
+    }
+}
+
+pub(crate) struct Tcp {
+    /// Bound server sockets
+    binds: IndexMap<SocketAddr, mpsc::UnboundedSender<(Syn, SocketAddr)>>,
+
+    /// Active stream sockets
+    sockets: IndexMap<SocketPair, StreamSocket>,
+}
+
+struct StreamSocket {
+    buf: IndexMap<u64, StreamData>,
+    next_send_seq: u64,
+    recv_seq: u64,
+    sender: mpsc::UnboundedSender<StreamData>,
+}
+
+impl StreamSocket {
+    fn new() -> (Self, mpsc::UnboundedReceiver<StreamData>) {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let sock = Self {
+            buf: IndexMap::new(),
+            next_send_seq: 1,
+            recv_seq: 0,
+            sender: tx,
+        };
+
+        (sock, rx)
+    }
+
+    fn assign_seq(&mut self) -> u64 {
+        let seq = self.next_send_seq;
+        self.next_send_seq += 1;
+        seq
+    }
+
+    // Buffer and re-order received segments by `seq` as the network may deliver
+    // them out of order.
+    fn buffer(&mut self, seq: u64, data: StreamData) {
+        assert!(
+            self.buf.insert(seq, data).is_none(),
+            "duplicate segment {}",
+            seq
+        );
+
+        while self.buf.contains_key(&(self.recv_seq + 1)) {
+            self.recv_seq += 1;
+
+            let data = self.buf.remove(&self.recv_seq).unwrap();
+            let _ = self.sender.send(data);
+        }
+    }
+}
+
+impl Tcp {
+    fn install() -> Self {
+        Self {
+            binds: IndexMap::new(),
+            sockets: IndexMap::new(),
+        }
+    }
+
+    fn is_port_assigned(&self, port: u16) -> bool {
+        self.binds.keys().any(|a| a.port() == port)
+            || self.sockets.keys().any(|a| a.local.port() == port)
+    }
+
+    pub(crate) fn bind(&mut self, addr: SocketAddr) -> io::Result<TcpListener> {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        if self.binds.insert(addr, tx).is_some() {
+            return Err(io::Error::new(io::ErrorKind::AddrInUse, addr.to_string()));
+        }
+
+        trace!(?addr, protocol = %"TCP", "Bind");
+
+        Ok(TcpListener::new(addr, rx))
+    }
+
+    pub(crate) fn new_stream(&mut self, pair: SocketPair) -> mpsc::UnboundedReceiver<StreamData> {
+        let (sock, rx) = StreamSocket::new();
+
+        assert!(
+            self.sockets.insert(pair, sock).is_none(),
+            "{:?} is already connected",
+            pair
+        );
+
+        rx
+    }
+
+    // Ideally, we could "write through" the tcp software, but this is necessary
+    // due to borrowing the world to access the mut host and for sending.
+    pub(crate) fn assing_send_seq(&mut self, pair: SocketPair) -> Option<u64> {
+        let sock = self.sockets.get_mut(&pair)?;
+        Some(sock.assign_seq())
+    }
+
+    fn receive_from_network(
+        &mut self,
+        src: SocketAddr,
+        dst: SocketAddr,
+        segment: Segment,
+    ) -> Option<Protocol> {
```

Comment:
i'm not convinced that this should be `Option`—maybe a Result instead?

Reply:
I could see a Result with an Err(Rst) reading nicely here.